radix-engine/src/kernel/call_frame.rs

```diff
@@ -356,16 +392,49 @@ impl<L: Clone> CallFrame<L> {
         }
 
         // Only allow move of `Global` and `DirectAccess` references
-        for node_id in message.copy_references {
-            if let Some(t) = from
-                .get_node_visibility(&node_id)
-                .can_be_reference_copied_to_frame()
-            {
+        for node_id in message.copy_global_references {
+            if from.get_node_visibility(&node_id).is_global() {
                 // Note that GLOBAL and DirectAccess references are mutually exclusive,
                 // so okay to overwrite
-                to.stable_references.insert(node_id, t);
+                to.stable_references
+                    .insert(node_id, StableReferenceType::Global);
+            } else {
+                return Err(PassMessageError::GlobalRefNotFound(node_id));
+            }
+        }
+
+        for node_id in message.copy_direct_access_references {
+            if from.get_node_visibility(&node_id).can_be_invoked(true) {
+                to.stable_references
+                    .insert(node_id, StableReferenceType::DirectAccess);
             } else {
-                return Err(PassMessageError::StableRefNotFound(node_id));
+                return Err(PassMessageError::DirectRefNotFound(node_id));
+            }
+        }
+
+        for node_id in message.copy_to_stable_transient_references {
+            if from.depth >= to.depth {
+                panic!("Transient references only supported for downstream calls.");
```

Comment:
Panic? Can it be triggered by clients?

Reply:
Nope, can never be triggered by clients as this is controlled by the system layer. You can see that `copy_to_stable_transient_references` is empty in `CallFrameMessage::from_output()`